starlette/routing.py

```diff
@@ -647,6 +651,19 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
 
         partial = None
 
+        if self.user_ctx is not None:
+            current_ctx = contextvars.copy_context()
+            for cvar in self.user_ctx:
+                if cvar in current_ctx:
+                    # not set by the user, skip to avoid modifying
+                    # vars used by event loops, servers, etc.
+                    continue
+                try:
+                    if cvar.get() != self.user_ctx.get(cvar):
+                        cvar.set(self.user_ctx.get(cvar))
+                except LookupError:
+                    cvar.set(self.user_ctx.get(cvar))
```

Comment:
I'm pretty sure this sort of copying back the context can be used to allow things run by `run_in_threadpool` to behave like if they were being called directly (which would solve issues FastAPI has with sync dependencies)